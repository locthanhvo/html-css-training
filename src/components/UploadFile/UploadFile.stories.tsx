import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import UploadFile from '.';

export default {
  title: 'Components/UploadFile',
  component: UploadFile,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof UploadFile>;

export const Default: Story = {
  args: {},
};
