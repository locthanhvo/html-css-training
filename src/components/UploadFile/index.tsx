import { HTMLAttributes, memo } from 'react';
import { Center, Flex, Input, Text } from '@chakra-ui/react';

// Icons
import { AttachmentIcon } from '@/components/common/Icons';

interface UploadFileProps {
  previewURLs?: string[];
  variant?: 'icon' | 'button';
  getRootProps?: (
    props?: HTMLAttributes<HTMLElement>,
  ) => HTMLAttributes<HTMLElement>;
  getInputProps?: (
    props?: HTMLAttributes<HTMLElement>,
  ) => HTMLAttributes<HTMLElement>;
}

const UploadFile = ({
  previewURLs = [],
  variant = 'icon',
  getInputProps = (props) => ({ ...props }),
  getRootProps = (props) => ({ ...props }),
}: UploadFileProps) => {
  return (
    <Center
      gap={3}
      w={variant === 'button' ? 168 : 'none'}
      h={8}
      border={variant === 'button' ? '2px solid' : 'none'}
      borderColor="lightGray"
      borderRadius="md"
      cursor="pointer"
      {...getRootProps()}
    >
      <Flex alignItems="center" justifyContent="center" gap={2}>
        {variant === 'icon' && <AttachmentIcon />}
        {variant === 'icon' && !!previewURLs.length && (
          <Text fontSize="sm" color="coolGray" textAlign="center">
            {previewURLs.length}
          </Text>
        )}
      </Flex>
      {variant === 'button' && (
        <Text fontSize="sm" color="primary" textAlign="center">
          Attachment
        </Text>
      )}
      <Input display="none" {...getInputProps()} />
    </Center>
  );
};

export default memo(UploadFile);
