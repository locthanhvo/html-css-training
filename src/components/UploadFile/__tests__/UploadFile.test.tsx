import { render, RenderResult } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import UploadFile from '..';

const mockGetInputProps = jest.fn();
const mockGetRootProps = jest.fn();

describe('UploadFile Component', () => {
  let renderResult: RenderResult;
  beforeEach(() => {
    renderResult = render(
      <BrowserRouter>
        <UploadFile
          getInputProps={mockGetInputProps}
          getRootProps={mockGetRootProps}
          previewURLs={[
            'https://images.unsplash.com/photo-1555041469-a586c61ea9bc',
          ]}
          variant="button"
        />
      </BrowserRouter>,
    );
  });
  it('should match snapshot', () => {
    const { container } = renderResult;
    expect(container).toMatchSnapshot();
  });

  it('should render icon upload', () => {
    const { getByText } = render(
      <BrowserRouter>
        <UploadFile
          getInputProps={(props) => ({ ...props })}
          getRootProps={(props) => ({ ...props })}
          previewURLs={[
            'https://images.unsplash.com/photo-1555041469-a586c61ea9bc',
          ]}
        />
      </BrowserRouter>,
    );

    expect(getByText('1')).toBeInTheDocument();
  });

  it('should call getRootProps and getInputProps with default props', () => {
    const { container } = render(
      <BrowserRouter>
        <UploadFile
          previewURLs={[
            'https://images.unsplash.com/photo-1555041469-a586c61ea9bc',
          ]}
        />
      </BrowserRouter>,
    );

    expect(container).toBeInTheDocument();
  });

  it('should render with previewURL empty ', () => {
    const { container } = render(
      <BrowserRouter>
        <UploadFile />
      </BrowserRouter>,
    );

    expect(container).toBeInTheDocument();
  });
});
