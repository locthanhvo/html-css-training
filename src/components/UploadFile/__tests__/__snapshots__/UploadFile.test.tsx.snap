// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadFile Component should match snapshot 1`] = `
<div>
  <div
    class="css-hmqloq"
  >
    <div
      class="css-w1wtgb"
    />
    <p
      class="chakra-text css-15xnlmf"
    >
      Attachment
    </p>
    <input
      class="chakra-input css-1hyfx7x"
    />
  </div>
</div>
`;
